prime-router/src/main/kotlin/Element.kt

```diff
@@ -759,6 +766,72 @@ data class Element(
         return codeValue?.code
     }
 
+    /**
+     * Determines if an element needs to use a mapper given the [elementValue].
+     * @return true if a mapper needs to be run
+     */
+    fun useMapper(elementValue: String): Boolean {
+        val overrideValue = mapperOverridesValue != null && mapperOverridesValue
+        return mapperRef != null && (overrideValue || elementValue.isBlank())
+    }
+
+    /**
+     * Determines if an element needs to use a default given the [elementValue].
+     * @return true if a default needs to be used
+     */
+    fun useDefault(elementValue: String): Boolean {
+        return elementValue.isBlank()
+    }
+
+    /**
+     * Determine the value for this element based on the schema configuration.  This function checks if a
+     * mapper needs to be run or if a default needs to be applied.
+     * @param allElementValues the values for all other elements.  Used for the mappers.
+     * @param schema the schema
+     * @param defaultOverrides element name and value pairs of defaults that override schema defaults
+     * @return a mutable set with the processed value or empty string
+     */
+    fun processValue(
+        allElementValues: Map<String, String>,
+        schema: Schema,
+        defaultOverrides: Map<String, String> = emptyMap()
+    ): String {
+        var retVal = if (allElementValues[name].isNullOrEmpty()) "" else allElementValues[name]!!
+        if (useMapper(retVal)) {
+            // This gets the requiredvalue names, then gets the value from mappedRows that has the data
+            val args = mapperArgs ?: emptyList()
+            val valueNames = mapperRef?.valueNames(this, args) ?: emptyList()
+            val valuesForMapper = valueNames.mapNotNull { elementName ->
+                val valueElement = schema.findElement(elementName)
+                if (valueElement != null && allElementValues.containsKey(elementName) &&
+                    !allElementValues[elementName].isNullOrEmpty()
+                ) {
+                    ElementAndValue(valueElement, allElementValues[elementName]!!)
+                } else {
+                    null
+                }
+            }
+            // Only overwrite an existing value if the mapper returns a string
+            val value = mapperRef?.apply(this, args, valuesForMapper)
+            if (!value.isNullOrBlank()) {
+                retVal = value
+            }
+        }
+
+        // Finally, add a default value or empty string to elements that still have a null value.
```

Comment:
Interesting.   The default applies _after_ the mapper.   I would have thought the default would apply _before_ the mapper, then the mapper might do further transforms on it (or on the external/incoming value if there is one).

But I haven't thought about this in detail.